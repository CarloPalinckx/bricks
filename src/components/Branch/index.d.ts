import { SFC, ReactNode } from 'react';
declare type PropsType = {
    condition: boolean;
    ifTrue(children: ReactNode): JSX.Element;
    ifFalse?(children: ReactNode): JSX.Element;
};
declare const Branch: SFC<PropsType>;
export default Branch;
export { PropsType };
