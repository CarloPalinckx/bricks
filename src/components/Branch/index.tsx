import { FunctionComponent, ReactNode, Children } from 'react';

type PropsType = {
    condition: boolean;
    ifTrue(children: ReactNode): JSX.Element;
    ifFalse?(children: ReactNode): JSX.Element;
};

const Branch: FunctionComponent<PropsType> = ({ condition, ifTrue, ifFalse, children }): JSX.Element | null => {
    if (condition) {
        return ifTrue(children);
    }

    if (ifFalse !== undefined) {
        return ifFalse(children);
    }

    if (Children.count(children) > 0) {
        return children as JSX.Element;
    }

    /* tslint:disable */
    return null;
    /* tslint:enable */
};

export default Branch;
export { PropsType };
