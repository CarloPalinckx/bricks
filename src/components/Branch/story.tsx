import { boolean } from '@storybook/addon-knobs/react';
import Branch from '.';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Text from '../Text';

storiesOf('Utilities/Branch', module).add('Default', () => {
    return (
        <Branch
            condition={boolean('condition', true)}
            ifTrue={(children): JSX.Element => <Text>This is shown when condition is true. {children}</Text>}
            ifFalse={(children): JSX.Element => <Text>And this will be shown when condition is false. {children}</Text>}
        >
            This line is shared whether the condition is true or not.
        </Branch>
    );
});
