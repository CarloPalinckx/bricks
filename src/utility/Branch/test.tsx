import React from 'react';
import { shallow } from 'enzyme';
import Branch from '.';
import toJson from 'enzyme-to-json';

describe('Branch', () => {
    it('should use true path', () => {
        const component = shallow(
            <Branch
                condition={true}
                ifTrue={(children): JSX.Element => (
                    <>
                        <span>true</span>
                        {children}
                    </>
                )}
            >
                <span>children</span>
            </Branch>,
        );

        expect(
            component
                .find('span')
                .first()
                .text(),
        ).toEqual('true');
        expect(
            component
                .find('span')
                .last()
                .text(),
        ).toEqual('children');
    });

    it('should use the false path', () => {
        const component = shallow(
            <Branch
                condition={false}
                ifTrue={(children): JSX.Element => (
                    <>
                        <span>true</span>
                        {children}
                    </>
                )}
                ifFalse={(children): JSX.Element => (
                    <>
                        <span>false</span>
                        {children}
                    </>
                )}
            >
                <span>children</span>
            </Branch>,
        );

        expect(
            component
                .find('span')
                .first()
                .text(),
        ).toEqual('false');
        expect(
            component
                .find('span')
                .last()
                .text(),
        ).toEqual('children');
    });

    it('should no-op when no false path is provided and the condition is false', () => {
        const component = shallow(
            <Branch
                condition={false}
                ifTrue={(children): JSX.Element => (
                    <>
                        <span>true</span>
                        {children}
                    </>
                )}
            >
                <span>children</span>
            </Branch>,
        );

        expect(component.find('span').text()).toEqual('children');
    });

    it('should not render anything when no components or false path are provided and the condition is false', () => {
        const component = shallow(<Branch condition={false} ifTrue={(): JSX.Element => <span>true</span>} />);

        expect(toJson(component)).toMatchSnapshot();
        expect(component.at(0).text()).toEqual('');
    });
});
