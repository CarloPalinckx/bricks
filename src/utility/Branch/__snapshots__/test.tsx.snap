// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Branch should not render anything when no components or false path are provided and the condition is false 1`] = `""`;
